File: construct-binary-tree-from-preorder-and-inorder-traversal.js
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No, did this in class.

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */

let index;
let map = {};
var buildTree = function(preorder, inorder) {
    if (preorder == null || preorder.length == 0) return null;
    index = 0;
    for (let i = 0; i < inorder.length; i++) {
        map[inorder[i]] = i;
    }
    
    return helper(preorder, inorder, 0, inorder.length - 1);
}

var helper = function (preorder, inorder, start, end) {
    // Base Case
    if (start > end) return null;
    
    // Logic
    let root = new TreeNode(preorder[index++]);
    
    // root index in my inorder
    let inIdx = map[root.val];
    root.left = helper(preorder, inorder, start, inIdx - 1);
    root.right = helper(preorder, inorder, inIdx + 1, end);
    
    return root;
}

File: Construct Binary_Tree_from_Preorder_and_Inorder_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Integer, Integer> map;
    int idx;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(preorder.length == 0){return null;} 
        
        map = new HashMap<>();
        for(int i = 0; i< inorder.length; i++){
            map.put(inorder[i], i);
        }
        idx = 0;
        return helper(preorder, inorder, 0 ,inorder.length-1);
    }
    
    private TreeNode helper(int[] preorder, int[] inorder, int start, int end){
        //base
        if(start > end) return null;
        //logic
        //find the root
        
        int rootIdx = map.get(preorder[idx]);
        TreeNode root = new TreeNode(preorder[idx]);
        idx++;
        System.out.println(idx);
        root.left = helper(preorder, inorder, start, rootIdx -1);
        root.right = helper(preorder, inorder, rootIdx + 1, end);
        return root;
    }
}

//TC : O(N)
//SC : O(H)
